app/client/ui/FormAPI.ts

```diff
@@ -90,68 +90,91 @@ export type FormOptionsSortOrder = 'default' | 'ascending' | 'descending';
+  createAttachments(options: CreateAttachmentOptions): Promise<number[]>;
 }
 
-interface GetFormCommonOptions {
-  vsId: number;
-}
-
-interface GetFormWithDocIdOptions extends GetFormCommonOptions {
+interface FormTargetWithDocId {
   docId: string;
 }
 
-interface GetFormWithShareKeyOptions extends GetFormCommonOptions {
+interface FormTargetWithShareKey {
   shareKey: string;
 }
 
-type GetFormOptions = GetFormWithDocIdOptions | GetFormWithShareKeyOptions;
+type FormTarget = FormTargetWithDocId | FormTargetWithShareKey;
+
+interface GetFormCommonOptions {
+  vsId: number;
+}
+
+type GetFormOptions = GetFormCommonOptions & FormTarget;
 
 interface CreateRecordCommonOptions {
   tableId: string;
   colValues: ColValues;
 }
 
-interface CreateRecordWithDocIdOptions extends CreateRecordCommonOptions {
-  docId: string;
-}
+type CreateRecordOptions = CreateRecordCommonOptions & FormTarget;
 
-interface CreateRecordWithShareKeyOptions extends CreateRecordCommonOptions {
-  shareKey: string;
+interface CreateAttachmentCommonOptions {
+  upload: File[];
 }
 
-type CreateRecordOptions = CreateRecordWithDocIdOptions | CreateRecordWithShareKeyOptions;
+type CreateAttachmentOptions = CreateAttachmentCommonOptions & FormTarget;
 
 export class FormAPIImpl extends BaseAPI implements FormAPI {
   constructor(private _homeUrl: string, options: IOptions = {}) {
     super(options);
   }
 
   public async getForm(options: GetFormOptions): Promise<Form> {
-    if ('docId' in options) {
-      const {docId, vsId} = options;
-      return this.requestJson(`${this._url}/api/docs/${docId}/forms/${vsId}`, {method: 'GET'});
-    } else {
-      const {shareKey, vsId} = options;
-      return this.requestJson(`${this._url}/api/s/${shareKey}/forms/${vsId}`, {method: 'GET'});
-    }
+    const {vsId} = options;
+    const url = this._getUrl(options, `forms/${vsId}`);
+    return this.requestJson(url.toString(), {method: 'GET'});
   }
 
   public async createRecord(options: CreateRecordOptions): Promise<void> {
-    if ('docId' in options) {
-      const {docId, tableId, colValues} = options;
-      return this.requestJson(`${this._url}/api/docs/${docId}/tables/${tableId}/records`, {
-        method: 'POST',
-        body: JSON.stringify({records: [{fields: colValues}]}),
-      });
+    const {tableId, colValues} = options;
+    const url = this._getUrl(options, `tables/${tableId}/records`);
+    return this.requestJson(url.toString(), {
+      method: 'POST',
+      body: JSON.stringify({records: [{fields: colValues}]}),
+    });
+  }
+
+  public async createAttachments(options: CreateAttachmentOptions): Promise<number[]> {
+    const url = this._getUrl(options, 'attachments');
+
+    options.upload = options.upload.filter(f => f.size > 0);
+    if (options.upload.length === 0) {
+      return [];
+    }
+
+    const formData = new FormData();
+    for (const file of options.upload) {
+      formData.append('upload', file);
+    }
```

Comment:
Nitpicking: I feel generally more securing to not alter the parameters, in Javascript / Typescript ownership is not clear (at the contrary of other languages [like Rust](https://doc.rust-lang.org/book/ch04-01-what-is-ownership.html)). It can lead to side-effect when the caller does not expect the arguments to be modified.

```suggestion
    const upload = options.upload.filter(f => f.size > 0);
    if (upload.length === 0) {
      return [];
    }

    const formData = new FormData();
    for (const file of upload) {
      formData.append('upload', file);
    }
```